### PR TITLE
Better color picker for settings page

### PR DESCRIPTION
**Pull request by GrahamSH-LLK**
This adds a web component color picker that has transparency.

image
image

Todos:
- [x] Clicking outside of color picker doesn't close it

### CODE REVIEW

**Comment by retronbv** [comment]
0063808C-A7A2-4DC8-B034-C3A9C0D614A2 could this use darkmode when it is on?

**Comment by GrahamSH-LLK** [comment]
Sure! I'll add it rn

**Comment by GrahamSH-LLK** [comment]
Dark and light mode added!

**Review by TheColaber — APPROVED** [review]

**Comment by mxmou** [comment]
Please don't do this. I'm working on customizable editor colors and letting the user configure transparency of every color would make it much harder to determine the appropriate text color for all elements.

Transparency can be a separate setting if necessary anyway.

**Comment by GrahamSH-LLK** [comment]
> Please don't do this. I'm working on customizable editor colors and letting the user configure transparency of every color would make it much harder to determine the appropriate text color for all elements.
> 
> Transparency can be a separate setting if necessary anyway.

Hmm. So should this be a "transparent-color" setting? Shouldn't users be able to pick any text color, even if it looks terrible? That's what SA is about....

**Comment by mxmou** [comment]
> Hmm. So should this be a "transparent-color" setting? Shouldn't users be able to pick any text color, even if it looks terrible? That's what SA is about....

Users aren't able to pick text colors by themselves because that would mean even more colors than [this](https://github.com/ScratchAddons/ScratchAddons/issues/1680#issuecomment-787109526). Instead, a userscript determines the appropriate text color (white or dark gray) for every background color, just like `scratchr2`, to make sure text is readable.

Making this a new setting type (and changing existing addons that would work with transparency, such as remix tree button) is probably the best solution. It might make sense to also change the normal color setting to use this library, just without transparency, for consistency.

File: webpages/settings/index.js [hunk 1]
```diff
@@ -1,13 +1,67 @@
...
+chrome.permissions.getAll(({ permissions }) => {
+  let grantedOptionalPermissions = permissions.filter((p) => browserLevelPermissions.includes(p));
+});
 let grantedOptionalPermissions = [];
-const updateGrantedPermissions = () =>
-  chrome.permissions.getAll(({ permissions }) => {
-    grantedOptionalPermissions = permissions.filter((p) => browserLevelPermissions.includes(p));
-  });
-updateGrantedPermissions();
+const updateGrantedPermissions = () => updateGrantedPermissions();
```
Comment (GrahamSH-LLK) [inline webpages/settings/index.js:64]: I'm not sure what happened here....

**Comment by WorldLanguages** [comment]
I agree on making transparency optin for addons

**Comment by GrahamSH-LLK** [comment]
I think using an input[type=color] will be fine for non-transparent items? cc @WorldLanguages @mxmou

**Comment by GrahamSH-LLK** [comment]
Non-transparent settings are added!

**Comment by GrahamSH-LLK** [comment]
image
vs
image

**Comment by Endermaster3690** [comment]
What about the scratch paint editor? but idk how it and #1768 will combine.

**Comment by GrahamSH-LLK** [comment]
These are completely different. The other one is in the editor, whereas this is for the settings page. They are completely compatible in the fact that they don't touch each other.

**Comment by GrahamSH-LLK** [comment]
@WorldLanguages Review?

**Comment by WorldLanguages** [comment]
1. Can't we use the color picker component for color settings that don't allow transparency? Doesn't the component provide a way to not allow the user to set transparency?
2. Unlike the browser color picker, the input value only updates after the color picker is closed. Can we change that?

**Comment by GrahamSH-LLK** [comment]
1. No.
2. Sure, but it'll be more resource intensive

File: webpages/settings/index.js [hunk 2]
```diff
@@ -76,7 +135,10 @@ const deserializeSettings = async (str, manifests, confirmElem) => {
     }
     addonSettings[addonId] = Object.assign({}, addonSettings[addonId], addonValue.settings);
   }
-  if (handleConfirmClicked) confirmElem.removeEventListener("click", handleConfirmClicked, { once: true });
+  if (handleConfirmClicked)
+    confirmElem.removeEventListener("click", handleConfirmClicked, {
+      once: true,
+    });
```
Comment (WorldLanguages) [inline webpages/settings/index.js:141]: Did something change here?

File: webpages/settings/index.js [hunk 3]
```diff
@@ -324,7 +365,9 @@ const vue = (window.vue = new Vue({
             : event.shiftKey
             ? false
             : newState;
-        chrome.runtime.sendMessage({ changeEnabledState: { addonId: addon._addonId, newState } });
+        chrome.runtime.sendMessage({
+          changeEnabledState: { addonId: addon._addonId, newState },
+        });
```
Comment (WorldLanguages) [inline webpages/settings/index.js:370]: Ditto

File: webpages/settings/index.js [hunk 4]
```diff
@@ -368,7 +411,10 @@ const vue = (window.vue = new Vue({
       setTimeout(() => {
         if (!settingId || this.addonSettings[addon._addonId][settingId] === value) {
           chrome.runtime.sendMessage({
-            changeAddonSettings: { addonId: addon._addonId, newSettings: this.addonSettings[addon._addonId] },
+            changeAddonSettings: {
+              addonId: addon._addonId,
+              newSettings: this.addonSettings[addon._addonId],
+            },
```
Comment (WorldLanguages) [inline webpages/settings/index.js:417]: lol

File: webpages/settings/index.js [hunk 5]
```diff
@@ -540,7 +586,15 @@ const vue = (window.vue = new Vue({
         this.isOpen = false;
       }
     },
+    closePickers(e) {
+      for (let child of this.$children) {
```
Comment (WorldLanguages) [inline webpages/settings/index.js:590]: Is `child` guaranteed to be a color picker component?

Comment (GrahamSH-LLK) [reply webpages/settings/index.js:590]: No. We just have no other components on the page.

**Comment by WorldLanguages** [comment]
3. Starting a click inside the component, but releasing the click outside the component, counts as "clicking outside".

**Comment by GrahamSH-LLK** [comment]
> 1. Starting a click inside the component, but releasing the click outside the component, counts as "clicking outside".

Is this really a problem? I don't really know how to fix that.

File: webpages/settings/index.js [hunk 6]
```diff
@@ -24,18 +24,25 @@ Vue.directive("click-outside", {
...
       isOpen: false,
       color: this.value,
       canCloseOutside: false,
+      formats: '',
     };
   },
   ready() {
-    self = this;
-    this.$els.pickr.addEventListener("change", (e) => {
+    self = this
```
Comment (WorldLanguages) [inline webpages/settings/index.js]: Is this creating a global variable?

Comment (WorldLanguages) [reply webpages/settings/index.js]: It is mutating window.self, can you use `const that = this` instead? (or directly use `this` if possible)

File: webpages/settings/index.js [hunk 7]
```diff
@@ -35,18 +35,17 @@ const ColorInput = Vue.extend({
...
       this.formats = "hex,hex8,rgb,hsv,hsl";
     }
-    console.log(this.no_alpha);
-    this.$els.pickr.addEventListener("change", (e) => {
-      self.color = "#" + e.detail.value;
-      if (self.value !== self.color) {
-        self.$parent.addonSettings[self.addon._addonId][self.setting.id] = "#" + self.$els.pickr.hex8;
-        self.$parent.updateSettings(self.addon, { wait: 250, settingId: self.setting.id });
+    this.$els.pickr.addEventListener('change', (e) => {
+      that.color = "#" + e.detail.value;
```
Comment (WorldLanguages) [inline webpages/settings/index.js:45]: You're using an arrow function, so `this === that // true` here, No need to use "that"

File: webpages/settings/index.js (hunk 5 above)
Comment (GrahamSH-LLK) [reply webpages/settings/index.js:606]: Also, no other components will have all of the correct methods it checks.

File: webpages/settings/index.html [hunk 8]
```diff
@@ -269,15 +286,24 @@ <h1>{{ msg("settings") }}</h1>
...
+                  ></picker>
+                  <picker
+                    v-if="setting.type === 'color' && !setting.allowTransparency"
+                    :value="addonSettings[addon._addonId][setting.id] || setting.default"
+                    :setting="setting"
+                    :addon="addon"
+                    @click=""
+                    no_alpha="true"
+                    v-click-outside="closePickers"
+                  ></picker>
```
Comment (WorldLanguages) [inline webpages/settings/index.html:306]: Make this a single element, with a dynamic `no_alpha` attribute.

File: libraries/CREDITS.md [hunk 9]
```diff
@@ -8,5 +8,7 @@ Third-party libraries included are:
 - [intl-messageformat 9.4.7](https://unpkg.com/intl-messageformat@9.4.7/intl-messageformat.umd.min.js) (MIT)
 - [tinycolor2 1.4.2](https://raw.githubusercontent.com/bgrins/TinyColor/1.4.2/dist/tinycolor-min.js) (MIT)
 - [Chart.js 2.9.4](https://unpkg.com/chart.js@2.9.4/dist/Chart.min.js) (MIT)
+- [color-picker-element 1.0.17](https://unpkg.com/color-picker-web-component@1.0.17/dist/color-picker-esm.js) (MIT)
```
Comment (WorldLanguages) [inline libraries/CREDITS.md:11]: Is it `color-picker-element` or `color-picker-web-component`?

File: libraries/license-info.json [hunk 10]
```diff
@@ -7,5 +7,6 @@
   "intl-messageformat": "BSD-3-Clause",
   "vue": "MIT",
   "tinycolor2": "MIT",
-  "chartjs": "MIT"
+  "chartjs": "MIT",
+  "color-picker-element": "MIT"
```
Comment (WorldLanguages) [inline libraries/license-info.json:11]: Change library name?

File: webpages/settings/index.html [hunk 11]
```diff
@@ -382,7 +408,9 @@ <h1 class="setting-heading">{{ msg("moreSettings") }}</h1>
             >
           </p>
           <p>
-            <a href="./licenses.html?libraries=intl-messageformat,vue" target="_blank">{{ msg("libraryCredits") }}</a>
+            <a href="./licenses.html?libraries=intl-messageformat,vue,color-picker-element" target="_blank"
```
Comment (WorldLanguages) [inline webpages/settings/index.html:411]: Change library name?

File: webpages/settings/index.js [hunk 12]
```diff
@@ -1,6 +1,89 @@
 import downloadBlob from "../../libraries/download-blob.js";
 const NEW_ADDONS = ["folders", "variable-manager", "scratchstats"];
 
+Vue.directive("click-outside", {
+  priority: 700,
+  bind() {
+    let self = this;
```
Comment (WorldLanguages) [inline webpages/settings/index.js:7]: Unnecessary?

File: webpages/settings/index.js [hunk 13]
```diff
@@ -1,6 +1,89 @@
...
+    this.event = function (event) {
+      console.log("emitting event");
+      self.vm.$emit(self.expression, event);
+    };
+    this.el.addEventListener("mousedown", this.stopProp);
+    document.body.addEventListener("mousedown", this.event);
+  },
+
+  unbind() {
+    console.log("unbind");
```
Comment (WorldLanguages) [inline webpages/settings/index.js:17]: Console logs can be removed

File: webpages/settings/index.js [hunk 14]
```diff
@@ -1,6 +1,89 @@
...
+    return {
+      isOpen: false,
+      color: this.value,
+      canCloseOutside: false,
+      formats: "",
+      opening: false,
+    };
+  },
+  ready() {
+    let that = this;
```
Comment (WorldLanguages) [inline webpages/settings/index.js:39]: "that" is unused

File: webpages/settings/index.js [hunk 15]
```diff
@@ -1,6 +1,89 @@
...
+
+      this.color = "#" + this.$els.pickr.hex8;
+      if (this.value !== this.color) {
+        this.$parent.addonSettings[addon._addonId][setting.id] = "#" + this.$els.pickr.hex8;
+        this.$parent.updateSettings(addon, { wait: 250, settingId: setting.id });
+      }
+      this.canCloseOutside = false;
+      setTimeout(() => {
+        this.canCloseOutside = true;
+      }, 100);
```
Comment (WorldLanguages) [inline webpages/settings/index.js:72]: A timeout of 0 should suffice.

**Pull request merged**